drivers/spi/spi_xec_qmspi_full_duplex.c

```diff
@@ -0,0 +1,710 @@
+	regs->cstm = cstm;
+	regs->mode_alt1 = malt1;
+	regs->tm_taps = taps[0];
+	regs->tm_taps_adj = taps[1];
+	regs->tm_taps_ctrl = taps[2];
+}
+
+static uint32_t qspi_source_clock_freq(void)
+{
+	struct pcr_regs const *pcr =
+		(struct pcr_regs *)(DT_REG_ADDR_BY_IDX(DT_NODELABEL(pcr), 0));
+
+	if (pcr->TURBO_CLK & MCHP_PCR_TURBO_CLK_96M) {
+		return MEC172X_QSPI_TURBO_SRC_CLOCK_HZ;
+	}
+	return MEC172X_QSPI_SRC_CLOCK_HZ;
+}
+
+/*
+ * Calculate QMSPI frequency divider register field value based upon
+ * the configured QMSPI input frequency: 48 or 96 MHz.
+ * The hardware divider is encoded as:
+ * 0 is divide by full divider range: 256 or 65536.
+ * Non-zero is divide by that value: 1 to 256 or 655356.
+ */
+static uint32_t qspi_encoded_fdiv(uint32_t freq_hz)
+{
+	uint32_t fdiv = 1;
+	uint32_t src_clk = qspi_source_clock_freq();
+
+	if (freq_hz < (src_clk / 256u)) {
+		fdiv = 0; /* HW fdiv = 0 is divide by 256 */
+	} else if (freq_hz < src_clk) {
+		/* truncated integer division may result in lower freq. */
+		fdiv = src_clk / freq_hz;
+	}
+
+	return fdiv;
+}
+
+/* Program QMSPI frequency divider field in mode register */
+static void qspi_set_frequency(struct qspi_regs *regs, uint32_t freq_hz)
+{
+	uint32_t fdiv, mode;
+
+	fdiv = qspi_encoded_fdiv(freq_hz);
+	mode = regs->mode & ~(XEC_QSPI_M_CLK_DIV_MASK);
+	mode |= ((fdiv << XEC_QSPI_M_CLK_DIV_POS) & XEC_QSPI_M_CLK_DIV_MASK);
+	regs->mode = mode;
+}
+
+static uint32_t qspi_get_frequency(struct qspi_regs *regs)
+{
+	uint32_t src_clk = qspi_source_clock_freq();
+	uint32_t fdiv = (regs->mode & XEC_QSPI_M_CLK_DIV_MASK)
+			>> XEC_QSPI_M_CLK_DIV_POS;
+
+	if (fdiv == 0) {
+		fdiv = 256u;
+	}
+
+	return (src_clk / fdiv);
+}
+
+/*
+ * SPI signalling mode: CPOL and CPHA
+ * CPOL = 0 is clock idle state is low, 1 is clock idle state is high
+ * CPHA = 0 Transmitter changes data on trailing of preceding clock cycle.
+ *          Receiver samples data on leading edge of clock cyle.
+ *        1 Transmitter changes data on leading edge of current clock cycle.
+ *          Receiver samples data on the trailing edge of clock cycle.
+ * SPI Mode nomenclature:
+ * Mode CPOL CPHA
+ *  0     0    0
+ *  1     0    1
+ *  2     1    0
+ *  3     1    1
+ * QMSPI has three controls, CPOL, CPHA for output and CPHA for input.
+ * SPI frequency < 48MHz
+ *	Mode 0: CPOL=0 CHPA=0 (CHPA_MISO=0 and CHPA_MOSI=0)
+ *	Mode 3: CPOL=1 CHPA=1 (CHPA_MISO=1 and CHPA_MOSI=1)
+ * Data sheet recommends when QMSPI set at >= 48MHz, sample and change data
+ * on the same edge.
+ *  Mode 0: CPOL=0 CPHA=0 (CHPA_MISO=1 and CHPA_MOSI=0)
+ *  Mode 3: CPOL=1 CPHA=1 (CHPA_MISO=0 and CHPA_MOSI=1)
+ */
+
+const uint8_t smode_tbl[4] = {
+	0x00u, 0x06u, 0x01u, 0x07u
+};
+
+const uint8_t smode48_tbl[4] = {
+	0x04u, 0x02u, 0x05u, 0x03u
+};
+
+static void qspi_set_signalling_mode(struct qspi_regs *regs, uint32_t smode)
+{
+	const uint8_t *ptbl;
+	uint32_t m;
+
+	ptbl = smode_tbl;
+	if (qspi_get_frequency(regs) >= 48000000u) {
```

Comment:
magic number, can this use any clock control api or macros?

Reply:
i.e. MEC152X_QSPI_SRC_CLOCK_HZ

Reply:
will change it to MHZ(48)
it has no reference to MEC15xx or MEC17xx specific